scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/GraphQLMerlinDatabaseService.java

```diff
@@ -897,6 +899,36 @@ public Collection<ResourceType> getResourceTypes(final PlanId planId)
     return allResourceTypes;
   }
 
+  @Override
+  @SuppressWarnings("unchecked")
+  public Map<ActivityDirectiveId, GoalId> getActivityIdToGoalIdMap(final PlanId planId)
+  throws MerlinServiceException, IOException
+  {
+    final var request = """
+        query {
+          activity_directive(where: {plan_id: {_eq: %d}}) {
+            id
+            source_scheduling_goal_id
+            source_scheduling_goal_invocation_id
+          }
+        }
+        """.formatted(planId.id());
+    final JsonObject response = postRequest(request).get();
+    final var data = response.getJsonObject("data");
+    final List<Map.Entry<ActivityDirectiveId, GoalId>> results = data.getJsonArray("activity_directive").getValuesAs(
+        $ -> {
+          final var obj = $.asJsonObject();
+          final var id = new ActivityDirectiveId(obj.getInt("id"));
+          if (obj.isNull("source_scheduling_goal_id")) return null;
+          final var source_goal = obj.getInt("source_scheduling_goal_id");
+          if (obj.isNull("source_scheduling_goal_invocation_id")) return entry(id, new GoalId(source_goal, -1, Optional.empty()));
+          final Long source_invocation = (long) obj.getInt("source_scheduling_goal_invocation_id");
+          return entry(id, new GoalId(source_goal, -1, Optional.of(source_invocation)));
+        }
+    );
+    return Map.ofEntries(results.stream().filter(Objects::nonNull).toArray(Map.Entry[]::new));
```

Comment:
Nonblocking: Is there a reason other than code style to build the map like this vs via declaring the map and using `data.getJsonArray("activity_directive").forEach()` to fill it out? Is it for the immutability of the Map returned by `Map.ofEntries`? No change is needed, I'm asking because the code stood out.

Reply:
Its just style. I tend to think in functional style when working with collections.